src/lib/patch.rs

```diff
@@ -0,0 +1,261 @@
+
+use std::clone::Clone;
+use std::collections::BTreeMap;
+
+use action::ActionType;
+
+// How close do two diffs have to be (in characters) to be merged into one patch?
+// We assume lines are 80 characters long.
+const DIST_THRESHOLD: i64 = 3 * 80;
+
+#[derive(Clone, Eq, Debug, PartialEq)]
+/// A patch represents a single change in the input files.
+///
+/// See the `Hunk` struct which aggregates a number of related patches.
+pub struct Patch {
+    action: ActionType,
+    start: usize,
+    length: usize,
+}
+
+impl Patch {
+    /// Create a new patch.
+    pub fn new(ty: ActionType, start: usize, length: usize) -> Patch {
+        Patch {
+            action: ty,
+            start: start,
+            length: length,
+        }
+    }
+
+    /// Character number where this patch begins in the original file.
+    ///
+    /// The lexeme affected by this patch lies in the interval `[start, end)`,
+    /// as a convenience for printing slices from the original file, i.e.
+    /// `println!("{}", &file[patch.start()..patch.end()])`.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// Character number where this patch ends in the original file.
+    ///
+    /// The lexeme affected by this patch lies in the interval `[start, end)`,
+    /// as a convenience for printing slices from the original file, i.e.
+    /// `println!("{}", &file[patch.start()..patch.end()])`.
+    pub fn end(&self) -> usize {
+        self.start + self.length
+    }
+
+    /// Action associated with this patch.
+    pub fn action(&self) -> &ActionType {
+        &self.action
+    }
+}
+
+#[derive(Clone, Eq, Debug, PartialEq)]
+/// A hunk is a group of patches close to one another in the original files.
+pub struct Hunk {
+    patches: Vec<Patch>,
+    start: usize,
+    length: usize,
+}
+
+impl Hunk {
+    /// Convert a patch into a hunk.
+    pub fn new(patch: Patch) -> Hunk {
+        Hunk {
+            patches: vec![patch.clone()],
+            start: patch.start,
+            length: patch.length,
+        }
+    }
+
+    fn end(&self) -> usize {
+        self.start + self.length
+    }
+
+    fn add_patch(&mut self, patch: Patch) {
+        if patch.start < self.start {
+            self.start = patch.start;
+        }
+        let p_end = patch.end();
+        if p_end > self.end() {
+            self.length += p_end - self.end();
+        }
+        self.patches.push(patch);
+    }
+
+    fn is_patch_close(&self, patch: &Patch) -> bool {
+        (self.start as i64 - patch.start as i64).abs() <= DIST_THRESHOLD
+    }
+
+    /// Given a character number, return type and length of any patches that
+    /// the character is involved in.
+    ///
+    /// The return value is the action type (`DELETE`, `INSERT`, `MOVE`,
+    /// `UPDATE`) and the number of characters from `ch` that are involved in
+    /// that patch.
+    pub fn get_action(&self, ch: usize) -> Option<(ActionType, usize)> {
+        if ch >= self.start && ch < self.end() {
```

Comment:
No, this is saying "we have some random character from an input file, is the character involved in a patch in this hunk?". The hunk is made up of a number of related patches that are "close" to one another, but the patches are not necessarily adjacent, so there may be gaps between them (the examples above all have one hunk, but there are definitely gaps between the patches!). If the diff output only showed the context of each hunk, we would be able to print out a header above each one, like git-diff does by default.